### ISSUE TITLE
`after` and `before` options treat certain semver version numbers as date

### ISSUE THREAD

**Issue by mischah** [ericmatthys/grunt-changelog#51]
tl;dr

Git Tags like `2.6.19` are treated as date (`2019-02-06`) :grimacing:

---

We are using Git tags which are equal to the version number within our `package.json` file to handle relases.

Therefore our settings for changelog Grunt task are the following to generate a simple changelog with sections for every version (release).

```
changelog: {
  release: {
    options: {
      fileHeader: '# Changelog',
      logArguments: [
        '--pretty=%h - %ad: %s',
        '--no-merges',
        '--date=short'
      ],
      after: '<%= pkpCopy.version %>',
      dest : 'CHANGELOG.md',
      insertType: 'prepend',
      template: '## Version <%= pkg.version %> ({{date}})\n\n{{> features}}',
      featureRegex: /^(.*)$/gim,
      partials: {
        features: '{{#if features}}{{#each features}}{{> feature}}{{/each}}{{else}}{{> empty}}{{/if}}\n',
        feature: '- {{{this}}} {{this.date}}\n'
      }
    }
  }
}
```

In one project we encountered the following error while running our release task:

```
FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - process out of memory
Abort trap: 6
```

Running Grunt in verbose mode showed that the git log params are weird:

```
git log --pretty=%h - %ad: %s --no-merges --date=short --after="2019-02-06T00:00:00+01:00" --before="2015-05-06T19:57:06+02:00“
```

Instead of:

```
git log 2.6.19..HEAD --pretty=%h - %ad: %s --no-merges --date=short
```

And that’s the reason of this bug:
We reached a version number that »looks« like a date :flushed:
Because `2.6.19` is interpreted as `2019-02-06`

That is because the changelog task is trying to be too clever.
The `after` and `before` options are taking date strings, commit SHAs and Git tags and need to decide the how to handle them.
See: https://github.com/ericmatthys/grunt-changelog#optionsafter

Here is the decision making process:
https://github.com/ericmatthys/grunt-changelog/blob/master/tasks/changelog.js#L42-L43

Moment.js is used to check if the string is a valid date. :bomb: :boom:

My idea how to fix that is to check if the string is a valid semver versionnumber before checking if it is a valid date.

**Comment by krnlde** [comment]
The main problem why this task is created still exists. It worked for 2 versions but now I get the error again:

```
Running "changelog:release" (changelog) task
FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - process out of memory
```

Unfortunately there's no error line specified. All I can say so far is that I want to make release `"version": "2.6.19"` and the node/iojs process peaked up to 1.2GB in RAM when doing the changelog thingy.

**Comment by krnlde** [comment]
I found out why, `git log` seems to like to page every output and sometimes has a `END` character which can be quit via `q`. To solve this you can add a `--no-pager` argument to the `git log` (continue [here](http://stackoverflow.com/a/2365076/622041) for further reading). It'll look for example like this:

```
git --no-pager log 2.6.17..HEAD --pretty="%h - %ad: %s" --no-merges --date=short
```

The `--no-pager` has to be **in front** of the log.

**Comment by mischah** [comment]
Just tried to reproduce that on my machine (running git version 2.3.4 on Mac OS 10.9.5) and I had no problem with that. So it might be a windows problem. 

Will add that to the Pull request #52 for the windows users out there :hankey:

**Comment by mischah** [comment]
Update. I can reproduce that problem with longer changelogs :flushed:

So thanks for reporting that one :+1: @krnlde

**Comment by krnlde** [comment]
Any news?

**Comment by mischah** [comment]
@krnlde 

Sorry for the replying late. Its merged and with v0.3.1 on npm since May 27th :octocat:

**Comment by krnlde** [comment]
Sweet. I'll pull and see if it still works.
